xenomint/state.go

```diff
@@ -0,0 +1,494 @@
+				NodeID:    "",
+				Timestamp: time.Now(),
+				RowCount:  0,
+				LogOffset: savepoint,
+			},
+		},
+	}
+	return
+}
+
+func (s *State) replay(req *types.Request, resp *types.Response) (err error) {
+	var (
+		ierr      error
+		savepoint uint64
+		query     = &QueryTracker{req: req, resp: resp}
+	)
+	s.Lock()
+	defer s.Unlock()
+	savepoint = s.getID()
+	if resp.Header.ResponseHeader.LogOffset != savepoint {
+		err = ErrQueryConflict
+		return
+	}
+	for i, v := range req.Payload.Queries {
+		if _, ierr = s.writeSingle(&v); ierr != nil {
+			err = errors.Wrapf(ierr, "execute at #%d failed", i)
+			s.rollbackTo(savepoint)
+			return
+		}
+	}
+	s.setSavepoint()
+	s.pool.enqueue(savepoint, query)
+	return
+}
+
+func (s *State) ReplayBlock(block *types.Block) (err error) {
+	var (
+		ierr   error
+		lastsp uint64 // Last savepoint
+	)
+	s.Lock()
+	defer s.Unlock()
+	for i, q := range block.QueryTxs {
+		var query = &QueryTracker{req: q.Request, resp: &types.Response{Header: *q.Response}}
+		lastsp = s.getID()
+		if q.Response.ResponseHeader.LogOffset > lastsp {
+			err = ErrMissingParent
+			return
+		}
+		// Match and skip already pooled query
+		if q.Response.ResponseHeader.LogOffset < lastsp {
+			if !s.pool.match(lastsp, q.Request) {
+				err = ErrQueryConflict
+				return
+			}
+			continue
+		}
+		// Replay query
+		for j, v := range q.Request.Payload.Queries {
+			if q.Request.Header.QueryType == types.ReadQuery {
+				continue
+			}
+			if q.Request.Header.QueryType != types.WriteQuery {
+				err = errors.Wrapf(ErrInvalidRequest, "replay block at %d:%d", i, j)
+				s.rollbackTo(lastsp)
+				return
+			}
+			if _, ierr = s.writeSingle(&v); ierr != nil {
+				err = errors.Wrapf(ierr, "execute at %d:%d failed", i, j)
+				s.rollbackTo(lastsp)
+				return
+			}
+		}
+		s.setSavepoint()
+		s.pool.enqueue(lastsp, query)
+	}
+	// Check if the current transaction is ok to commit
+	if s.pool.matchLast(lastsp) {
+		if err = s.unc.Commit(); err != nil {
+			// FATAL ERROR
+			return
+		}
+		if s.unc, err = s.strg.Writer().Begin(); err != nil {
+			// FATAL ERROR
+			return
+		}
+		s.setNextTxID()
+	} else {
+		// Set commit point only, transaction is not actually committed. This commit point will be
+		// used on exiting.
+		s.setCommitPoint()
+	}
+	s.setSavepoint()
+	// Truncate pooled queries
+	s.pool.truncate(lastsp)
+	return
+}
+
+func (s *State) commit(out chan<- command) (err error) {
+	var queries []*QueryTracker
+	s.Lock()
+	defer s.Unlock()
+	if err = s.unc.Commit(); err != nil {
+		return
+	}
+	if s.unc, err = s.strg.Writer().Begin(); err != nil {
+		return
+	}
+	s.setNextTxID()
+	s.setSavepoint()
+	queries = s.pool.queries
+	s.pool = newPool()
+	// NOTE(leventeliu): Send commit request to the outcoming command queue within locking scope to
+	// prevent any extra writing before this commit.
+	if out != nil {
+		out <- &commitRequest{queries: queries}
+	}
+	return
+}
+
+func (s *State) CommitEx() (queries []*QueryTracker, err error) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.CommitEx should have comment or be unexported